File: library/src/main/java/io/github/anderscheow/validator/Validation.java
package io.github.anderscheow.validator;

import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;

import java.util.ArrayList;
import java.util.List;

import io.github.anderscheow.validator.conditions.Condition;
import io.github.anderscheow.validator.rules.BaseRule;

public class Validation {

    private TextInputLayout textInputLayout;

    private List<BaseRule> baseRules;
    private List<BaseRule> andRules;
    private List<BaseRule> orRules;

    private List<Condition> conditions;

    public Validation(@NonNull TextInputLayout textInputLayout) {
        this.textInputLayout = textInputLayout;

        this.baseRules = new ArrayList<>();
        this.andRules = new ArrayList<>();
        this.orRules = new ArrayList<>();
        this.conditions = new ArrayList<>();
    }

    /**
     * @deprecated Use {@link #and(BaseRule),
     *                  {@link #or(BaseRule) or
     *                  {@link #add(Condition)}}} )} instead.
     */
    @Deprecated
    public Validation addRule(@NonNull BaseRule baseRule) {
        baseRules.add(baseRule);
        return this;
    }

    public Validation and(@NonNull BaseRule baseRule) {
        andRules.add(baseRule);
        return this;
    }

    public Validation or(@NonNull BaseRule baseRule) {
        orRules.add(baseRule);
        return this;
    }

    public Validation add(@NonNull Condition condition) {
        conditions.add(condition);
        return this;
    }

    TextInputLayout getTextInputLayout() {
        return textInputLayout;
    }

    List<BaseRule> getBaseRules() {
        return baseRules;
    }

    List<BaseRule> getAndRules() {
        return andRules;
    }

    List<BaseRule> getOrRules() {
        return orRules;
    }

    List<Condition> getConditions() {
        return conditions;
    }
}

File: library/src/main/java/io/github/anderscheow/validator/Validation.kt
@file:Suppress("UNUSED")

package io.github.anderscheow.validator

import android.content.Context
import com.google.android.material.textfield.TextInputLayout
import io.github.anderscheow.validator.conditions.Condition
import io.github.anderscheow.validator.conditions.ConditionsBuilder
import io.github.anderscheow.validator.interfaces.ErrorImpl
import io.github.anderscheow.validator.rules.Rule
import io.github.anderscheow.validator.rules.RulesBuilder
import java.util.*

class Validation(val textInputLayout: TextInputLayout) {

    val rules = arrayListOf<Rule>()
    val conditions = arrayListOf<Condition>()

    fun add(rule: Rule): Validation {
        rules.add(rule)
        return this
    }

    fun add(condition: Condition): Validation {
        conditions.add(condition)
        return this
    }

    fun setError(context: Context, errorImpl: ErrorImpl, errors: ArrayList<String>) {
        val error = errorImpl.getErrorMessage(context)
        textInputLayout.isErrorEnabled = true
        errors.add(error)
        textInputLayout.error = error
    }
}

class ValidationBuilder {
    val ruleList = arrayListOf<Rule>()
    val conditionList = arrayListOf<Condition>()

    fun rules(init: RulesBuilder.() -> Unit) {
        ruleList.addAll(RulesBuilder().apply(init).ruleList)
    }

    fun conditions(init: ConditionsBuilder.() -> Unit) {
        conditionList.addAll(ConditionsBuilder().apply(init).conditionList)
    }
}

fun validation(textInputLayout: TextInputLayout, init: ValidationBuilder.() -> Unit): Validation {
    val validation = ValidationBuilder()
    validation.init()
    return Validation(textInputLayout).apply {
        this.rules.addAll(validation.ruleList)
        this.conditions.addAll(validation.conditionList)
    }
}